### PR TITLE
[RFC] Channel permissions

### PR DESCRIPTION
**Pull request by fowczarek**
The goal is to be able to limit staff users to only having access to certain channels. This should be configurable in the same way we configure permissions, all granted permissions would only work for selected channels.

## TODO list

- [x] Update `Group` model
- [x] Update `User` and `Group` types
- [x] Update permission group mutations: `PermissionGroupCreate`, `PermissionGroupUpdate`
- [ ] Update permission checking for permissions affected by this changes - check also if user has the access for the specific channel
  - [x] Order
  - [ ] Channel (post MVP)
  - [ ] Discount (post MVP)
  - [ ] Shipping (post MVP)
  - [ ] Products (post MVP)
  - [ ] Checkouts (post MVP)
- [x] Add OIDC support
- [ ] Docs update (**In review**)
  - [x] Upgrade guide
  - [x] Feature explanation

## Acceptance Criteria
- As a default, all staff users should have access to all channels
- The staff user should be able to specify the group channels during group creation and update
- When the staff user is in at least one group that has access to all channels, the user should see the items from all channels
- When the staff user has restricted access to a specific channel, the user should see the items only from a specific channel for queries and fields that requires the following permissions:
   - MANAGE_ORDERS
   - MANAGE_DISCOUNTS
   - MANAGE_PRODUCTS
   - MANAGE_SHIPPING
   - MANAGE_CHECKOUTS
- When the user is trying to fetch the items from channel which he doesn't have access to, the permission denied error is raisde
- The app should see the items from all channels.
- The user can careate/update the permission group only with the channels that are in his scope
- When the OpenID plugin is active, the new default group with access to all channels should be created and all staff users are assigned to this group. The default group name will be set in plugin configuration
- There is an option in the OpenID plugin to define staff user domains and default staff groups to which the new staff users should be assigned.
- The user logged with the OpenID is considered as staff when the email domain matches one of the domains from staff user domains, or the flag `use_scope_permissions` is set on the plugin and the user has some permissions assigned or the user is marked as staff.
- Feature support zero-downtime approach 

# 1. General Assumptions
- Staff users should be able to assign at least one channel to the permission group. 
- Staff users should be able to assign all channels to the permission group (including newly created channels).
- Staff users can change channels assigned to the permission group only when this staff user has access to modify this group and has access to all channels. 
- If the staff user is a member of many permissions groups, Saleor sums permissions and channels from all groups when checking that user should receive data.
    >  ⚠️⚠️⚠️  E.g.:
    >   **Group A:**
    >     - permissions: MANAGE_ORDERS
    >     - channels: channel_EU
    >   **Group B:**
    >     - permissions: MANAGE_PRODUCTS
    >     - channels: All
    >   **When the user is assigned to Group A and B, the user will have:**
    >     - `permissions`: MANAGE_ORDERS and MANAGE_PRODUCTS
    >     - `accessibleChannels`: All
- Apps permiss ions remain unchanged. Any app permission work for data in each channel. 
- The permission group is able to view and modify data only for channels that are assigned. Permissions affected by those changes(Not listed permissions are not affected by those changes.):
    - MANAGE_DISCOUNTS
    The staff user will be able to see all sales and vouchers but they can’t see and modify channel-specific data for restricted channels.
    - MANAGE_PRODUCTS
    The staff user will be able to see all products, variants, and collections but they can’t see and modify channel-specific data for restricted channels. Channel-specific data.
    - MANAGE_SHIPPING
    The staff user will be able to see all shipping methods but they can’t see and modify channel-specific data for restricted channels. Channel-specific data.
    - MANAGE_ORDERS
    The staff user will be able to see only orders from assigned channels.
    The staff user can create draft orders only from assigned channels.
    - MANAGE_CHECKOUTS
    The staff user will be able to see only checkout from assigned channels.

# 2. Database changes
Extend `Group` model to contain a list of assigned channels. 
Currently we use `django.contrib.auth.models.Group` we should extend this model by:
```python
class Group(models.Model):
	…
	restictedAccessToChannel = models.Boolean(default=False)
	channels = models.ManyToManyField(Channel)
```

In data migration, we assign channels without any restriction. 

# 2. API changes

### 2.1.1 User type
```graphql
type User { 
	…
	accessibleChannels: [Channel!]
	restictedAccessToChannel: Boolean!
}
```
- `accessibleChannels` - This field returns channels the user has to access through the assigned permissions group. If users don't have restricted channels this field returns all channels. 
- `restictedAccessToChannel` - this field returns `False` when the user has access to all channels(Even future created). If this field return `True` that means that the user has access only to the assigned channel from the field `accessibleChannels`.

### 2.1.2 Group type
```graphql
type Group {
	…
	accessibleChannels: [Chennel!]
	restictedAccessToChannel: Boolean!
}
```
- `accessibleChannels` - This field returns channels the group has to access. If the group doesn't have restricted channels this field returns all channels. 
- `restictedAccessToChannel` - this field returns `False` when the group has access to all channels(Even future created). If this field return `True` that means that the user has access only to the assigned channel from the field `accessibleChannels`.

### 2.1.3 Types with channel listings.
In all `channelListings` fields in types `ShippingMethodType`, `Product`, `ProductVariant`, `Collection`, `Sale`, `Voucher`,   we add information about this field returns only `ChannelListings` for channels available to the user. 

## 2.2 Extend mutations

> :warning: Warning :warning: 
> Staff users can change channels assigned to the permission group only when this staff user has access to modify this group and has access to all channels. 

### 2.2.1 Extend `PermissionGroupCreateInput` 
```graphql
input PermissionGroupCreateInput {
	…
	restictedAccessToChannel: Boolean!
	addChannels: [ID!]
}
```
- `restictAccesToChannel` - flag decided that those `PermissionGroup` should have access to all channels. If that flag is set on `False` then `addChannels` is skipped in assigning channels to the `PermissionGroup`. 
- `addChannels` - List of channels that should be assigned to `PermissionGroup`. This field is skipped when `allChannels` is set on `True`

### 2.2.2 Extend `PermissionGroupUpdateInput`

```graphql
input PermissionGroupUpdateInput {
	…
	restictedAccessToChannel: Boolean!
	addChannels: [ID!]
	removeChannels: [ID!]
}
```
- `restictedAccessToChannel` - flag decided that those `PermissionGroup` should have access to all channels. If that flag is set on `False` then `addChannels` and `removeChannels` are skipped in assigning channels to the `PermissionGroup`. 
When staff users mark the group as a group with access to all channels Saleor clears currently assigned channels. 
- `addChannels` - List of channels that should be assigned to `PermissionGroup`. This field is skipped when `allChannels` is set on `True`
- `removeChannels` -  List of channels that should be assigned to `PermissionGroup`. This field is skipped when `allChannels` is set on `True`

> :warning: Warning :warning: 
>The same channel id can’t appear in both lists(`addChannels` and `removeChannels`).

## 2.3 Behavior changes

- ~~`order`, `orderByToken`~~, `orders`, and `draftOrders` queriers - Any order query now checks if the requestor has assigned order permission and has access to the order’s channel.  
- `checkout`, `checkouts`, and `checkoutLines` queries - Any checkout query now checks if the requestor has assigned checkout permission and has access to the checkout’s channel.  
- Staff use could remove `ShippingMethodType`, `Product`, `ProductVariant`, `Collection`, `Sale`, `Voucher` only if have access to all channels for which this object is assigned(Object has created any Channel Listing in the channel for which admin don’t have access).
- Staff users could remove only those channel listings for channels to which they have access.

# 3. OIDC changes
## 3.1 Allows configuring staff user's default permission group via OIDC settings
We want to allow the OIDC plugin to assign newly created staff users to the chosen permission group. To be able to define if the user should be marked as a staff user we want to introduce a domain whitelist. 
Add new fields OIDC settings:
- `Staff users domains`: In this wield we store a list of domains separated by `,`(Store them as the lower case). If the email newly created user will be matched to any of these domains then the user will be marked as a staff user instead of a customer. 
- `Default group name for new staff users`: When any user will be created as staff then will be assigned to the permission group matching to group from this field. If the provided group doesn't exist Saleor we create a such group without any assigned permissions and channels.
> :warning: Warning :warning: 
> If `Default group name for new staff users` will be empty and user domain will be matched to any `Staff users domains`. Then Saleor creates a new staff user without any permission group assigned. 

## 3.2 Flow of new login process via OIDC:
```mermaid
graph TD
 A[User login via OIDC]
 A --> B{Check if the user has Saleor account}
 B -->|Yes| C[The user is successfully logged in]
 B -->|No| D{Check if the email domain match to<br> domain from OIDC settings}
 D -->|Yes| E[Create staff user in Saleor]
 D -->|No| C
 E --> F{Default permission group set in OIDC settings}
 F -->|Yes| G[Assing newly created staff user to default group<br> If the group does not exist create them without<br> any assigned permissions and channels]
 G -->C
 F -->|No| C
```

### 3.3.1 Migration process 
To do not break the currently working environments with RBAC, in the migration process we should create a permission group without any permission and all channels assigned to this group. Assign all staff users to this group. This migration should execute code only if OIDC is turned on. 

### 3.3.2 Logged-in user permission should be the union of permissions from OIDC and from the Saleor groups.
In pre Channel permissions we treat OIDC as a source of truth about user permission. In a new way, we want to create a union of permissions from OIDC and Saleor groups. 
After this change, OIDC will be provided only permissions and Saleor groups provide permissions and channel restrictions. 

# 4. Migration guide
- [ ] Add info about the new group `all_channels` and how it should be handled.
- [ ] Warning about that we implement 3.3.2

# Impact

* [ ] New migrations
* [ ] New/Updated API fields or mutations
* [ ] Deprecated API fields or mutations
* [ ] Removed API types, fields, or mutations
* [ ] Documentation needs to be updated

# Pull Request Checklist

* [ ] Privileged queries and mutations are guarded by proper permission checks
* [ ] Database queries are optimized and the number of queries is constant
* [ ] Database migration files are up to date
* [ ] The changes are tested
* [ ] GraphQL schema and type definitions are up to date
* [ ] Changes are mentioned in the changelog

### CODE REVIEW

**Comment by patrys** [comment]
Task linked: [SALEOR-7880 Channel permissions](https://app.clickup.com/t/2549495/SALEOR-7880)

**Comment by lkostrowski** [comment]
Hey @fowczarek 

I have 2 small inputs on public API changes:

- `restictAccesToChannel` - there is a typo - should be "access"
- `restictAccesToChannel` - about "restrict" verb conjugation. I understand this flag returns some existing state, that access was restricted and it's restricted now. So I think it should be `restricted` form. Present simple form ("restrict") suggestss that user can restrict something or perform any action/command

To sum up this part, I'd change this field to `restictedAccessToChannel` and consider this reasoning in future api changes

**Comment by IKarbowiak** [comment]
> In data migration, we assign channels without any restriction.

As I understand, all groups will have access to all channels?

> PermissionGroupUpdateInput

What about the situation when the Group has some channels assigned and `restictedAccessToChannel` set to `True`, and the value of `restictedAccessToChannel` is changed to `False` in the mutation? In such case, we will be removing all relations between this group and channels?

**Comment by fowczarek** [comment]
@IKarbowiak 
> As I understand, all groups will have access to all channels?

 Yes
 

> What about the situation when the Group has some channels assigned and restictedAccessToChannel set to True, and the value of restictedAccessToChannel is changed to False in the mutation? In such case, we will be removing all relations between this group and channels?

Yes, we want to clear it.

**Review by IKarbowiak — APPROVED** [review]

**Comment by korycins** [comment]
@fowczarek I think that we already talked about this, but not sure what we agree and I don't see a section dedicated to this.
What about an accessToken that is created by calling `apps.accessToken`. It contains shared permissions between user & app.  Like the below example. Staff user has all permissions, app has  manage_orders and handle_taxes. The permissions that are defined in token are (app.permissions & user.permissions).  It is something that should worth to mention in RFC
```
{
  "iat": 1662114083,
  "owner": "saleor",
  "exp": 1662114383,
  "token": "nVi3brvyeFIa",
  "email": "admin@example.com",
  "type": "thirdparty",
  "user_id": "VXNlcjox",
  "is_staff": true,
  "app": "QXBwOjE0",
  "permissions": [
    "MANAGE_ORDERS",
    "HANDLE_TAXES"
  ],
  "user_permissions": [
    "IMPERSONATE_USER",
    "MANAGE_DISCOUNTS",
    "MANAGE_PAGES",
    "MANAGE_SETTINGS",
    "MANAGE_MENUS",
    "MANAGE_PAGE_TYPES_AND_ATTRIBUTES",
    "HANDLE_PAYMENTS",
    "HANDLE_CHECKOUTS",
    "HANDLE_TAXES",
    "MANAGE_OBSERVABILITY",
    "MANAGE_PRODUCT_TYPES_AND_ATTRIBUTES",
    "MANAGE_CHANNELS",
    "MANAGE_USERS",
    "MANAGE_STAFF",
    "MANAGE_CHECKOUTS",
    "MANAGE_PRODUCTS",
    "MANAGE_PLUGINS",
    "MANAGE_APPS",
    "MANAGE_ORDERS",
    "MANAGE_TRANSLATIONS",
    "MANAGE_SHIPPING",
    "MANAGE_GIFT_CARD"
  ]
}
```

**Comment by fowczarek** [comment]
@korycins Apps have access to all channels. Only permission groups could have restricted access to channels. The mutual part of available channels will always be equal to the list of available channels for the user

**Review by szczecha — APPROVED** [review]

**Review by maarcingebala — APPROVED** [review]
